src/main/kotlin/com/xtyrrell/maxonbank/query/AccountProjector.kt

```diff
@@ -0,0 +1,95 @@
+package com.xtyrrell.maxonbank.query
+
+import com.xtyrrell.maxonbank.coreapi.*
+import org.axonframework.eventhandling.EventHandler
+import org.axonframework.queryhandling.QueryHandler
+import org.springframework.stereotype.Component
+import java.util.*
+
+@Component
+class AccountProjector(private val repository: AccountViewRepository) {
+
+    // An event-sourced approach would store keep a local variable `balance`
+    // here
+    //private val balance = 0
+    // Instead, we're storing state in a DB (including balance and a bunch of other things).
+    // Our approach is more memory efficient (for example, we don't need to keep in memory a list of all transactions,
+    // which is unbounded in size. So it's clearly better to store that in a DB.
+    // We could consider a hybrid approach where some data (like unbounded lists of transactions) are kept in the DB, while
+    // other data (like balances) are event-sourced. However, I don't see any clear benefits to this, so we're instead
+    // keeping all data in the DB.
+
+    @EventHandler
+    fun on(event: AccountOpenedEvent) {
+        val accountView = AccountView(event.accountId, 0)
+        repository.save(accountView)
+    }
+
+    @EventHandler
+    fun on(event: FundsDepositedEvent) {
+        val accountView: AccountView = repository.findById(event.accountId).get()
+
+        val entryId = UUID.randomUUID()
+        val entry = LedgerEntryView(accountView, entryId, LedgerEntryType.DEPOSIT, event.amount)
+        accountView.addLedgerEntry(entry)
+
+        accountView.balance += event.amount
```

Comment:
Well spotted! That's certainly worth some research. I can imagine it might be easier to keep all state in the database and manage race conditions through the concurrency mechanisms JPA provides.

Reply:
After thinking about this more:

There would be a race condition if
- two different JVMs each have their own AccountProjector instance handling a given `aggregateId`, and one instance receives the first FundsDepositedEvent and the other instance receives the other FundsDepositedEvent;
- or, if one JVM has two AccountProjector instances handling a given `aggregateId`, and one event is sent to each;
- or, if one JVM has one AccountProjector instance handling a given `aggregateId`, and both events are routed to it, but it doesn't process events atomically and so the first half of the FundsDepositedEvent event handler method is run for the first event, then before that can finish, the entire event handler is run for the second event, and only then the event handler for the first event finishes.

The last option seems unlikely. As far as I understand, Axon guarantees that there will always only be one instance for a given `aggregateId`, even in a distributed environment with multiple machines.

Definitely worth more research but so far it seems like this should be fine.